Retry if we get a reset during upload

philwills opened the pull request:
We've seen this happen a few times and it's always worked when re-run, so try to shortcut that.

Commit:
Retry if we get a reset during upload

We've seen this happen a few times and it's always worked when re-run, so try to shortcut that.

philwills commented:
Haven't been able to reproduce failure, so would appreciate a second pair of eyes 

cc @adamnfish @obrienm

johnduffell commented:
fair enough but I like the idea of updating the library version first as retrying things (aka non determinism) always makes me nervous! :+1:

adamnfish commented on s3-plugin-server/src/main/scala/com/gu/teamcity/ArtifactUploader.scala:
```diff
@@ -31,13 +33,21 @@ class ArtifactUploader(config: S3ConfigManager, s3: S3) extends BuildServerAdapt
+              case e: ResetException => {
+                runningBuild.addBuildMessage(new BuildMessage1(DefaultMessagesInfo.SOURCE_ID,
+                  DefaultMessagesInfo.MSG_BUILD_PROBLEM, Status.WARNING, new Date,
+                  s"Retrying artifact upload after error: ${e.getMessage}"))
```
`e.getMessage` isn't always amazingly useful feedback. I'm not familiar with the style here but would it be sensible to at least log the full error so more details are available?

philwills replied:
This is reporting back in the UI, rather than logging. Given that we've already established it's a reset at this point, I thought we'd probably gain little with more than the message.

adamnfish commented on s3-plugin-server/src/main/scala/com/gu/teamcity/ArtifactUploader.scala:
```diff
@@ -31,13 +33,21 @@ class ArtifactUploader(config: S3ConfigManager, s3: S3) extends BuildServerAdapt
+                  s"Retrying artifact upload after error: ${e.getMessage}"))
+
+                s3.upload(config.artifactBucket, runningBuild, buildArtifact.getName,
+                  buildArtifact.getInputStream, buildArtifact.getSize).map(_ => Continuation.CONTINUE)
```
It looks like we throw away the error if it fails in this block, which doesn't seem ideal?

philwills replied:
I figure if it's already failed once, and then fails again, there's probably little we're going to glean from it.

philwills replied:
Having said that, it's all very inelegant and I'm open to suggestion.

Commit:
Tidy up based on feedback from @adamnfish

Commit:
Pull upload function up a scope for clarity

Commit:
Tidy

Commit:
Don't need to pass around partial functions

Commit:
Throw an exception on failure

This is horrible, but makes the reported messages for the build more meaningful. Will take another look at a better approach.

philwills commented:
Turns out the Amazon client has retries which are configurable, so going to do that instead.

atward commented:
I'm getting this error in production. Is there a workaround?

philwills commented:
@atward Yes, #5 and #6 have changed the way we do the upload and that has removed the issue for us, but I forgot to push a release. I've now done that, so hopefully that will help.

atward commented:
Thanks. I'll retry with 1.4.0+ and report back.
